features/cloud.py

```diff
@@ -0,0 +1,373 @@
+                self.az_client_secret,
+                self.az_tenant_id,
+                self.az_subscription_id,
+            ]
+        )
+
+    @property
+    def api(self):
+        """Return the api used to interact with the cloud provider."""
+        if self._api is None:
+            self._api = pycloudlib.Azure(
+                tag=self.tag,
+                client_id=self.az_client_id,
+                client_secret=self.az_client_secret,
+                tenant_id=self.az_tenant_id,
+                subscription_id=self.az_subscription_id,
+            )
+
+        return self._api
+
+    def get_instance_id(self, instance):
+        """Return the instance identifier.
+
+        :param instance:
+            An instance created on the cloud provider
+        """
+        # For Azure, the API identifier uses the instance name
+        # instead of the instance id
+        return instance.name
+
+    def menage_ssh_keys(self, private_key_path=None):
+        """Create and manage ssh key pairs to be used in the cloud provider.
+
+        :param private_key_path:
+            Location of the private key path to use. If None, the location
+            will be a default location.
+        """
+        if not os.path.exists(self.AZURE_PUB_KEY_FILE):
+            if "uaclient-integration" in self.api.list_keys():
+                self.api.delete_key("uaclient-integration")
+            pub_key, priv_key = self.api.create_key_pair(
+                key_name="uaclient-integration"
+            )
+
+            with open(self.AZURE_PUB_KEY_FILE, "w") as stream:
+                stream.write(pub_key)
+
+            with open(self.AZURE_PRIV_KEY_FILE, "w") as stream:
+                stream.write(priv_key)
+
+            os.chmod(self.AZURE_PUB_KEY_FILE, 0o600)
+            os.chmod(self.AZURE_PRIV_KEY_FILE, 0o600)
+
+        self.api.use_key(
+            self.AZURE_PUB_KEY_FILE,
+            self.AZURE_PRIV_KEY_FILE,
+            "uaclient-integration",
+        )
+
+    def _create_instance(
+        self,
+        series: str,
+        image_name: "Optional[str]" = None,
+        user_data: "Optional[str]" = None,
+    ) -> pycloudlib.instance:
+        """Launch an instance for on the cloud provider.
+
+        :param series:
+            The ubuntu release to be used when creating an instance. We will
+            create an image based on this value if the used does not provide
+            a image_name value
+        :param image_name:
+            The name of the image to be used when creating the instance
+        :param user_data:
+            The user data to be passed when creating the instance
+
+        :returns:
+            An AWS cloud provider instance
+        """
+        if not image_name:
+            if not series:
+                raise ValueError(
+                    "Must provide either series or image_name to launch azure"
+                )
+            with open("features/azure-ids.yaml", "r") as stream:
```

Comment:
Given that you have pulled the pro image choice logic into the Cloud subclasses, I think we might also need to add a machine_type param to the __init__ of these Cloud instances so that we can determine if "generic.azure" or "generic.aws" machine_type is provided. Then if "pro" not in machine_type we can source azure-ids.yaml, otherwise we call cloud.released_image(series) for the non-pro machine_types